src/rfxencode_diff_rlgr1.c

```diff
@@ -0,0 +1,281 @@
+    while (bit_count >= 8) \
+    { \
+        bit_count -= 8; \
+        *cdata = bits >> bit_count; \
+        cdata++; \
+    } \
+} while (0)
+
+/*CodeGR(krp, lmag);*/ /* output GR code for (mag - 1) */
+#define CodeGR(_krp, _lmag) do { \
+    int lkr = _krp >> 3; \
+    /* unary part of GR code */ \
+    int lvk = _lmag >> lkr; \
+    /*OutputBit(lvk, 1);*/ \
+    int llvk = lvk; \
+    while (llvk >= 8) \
+    { \
+        bits <<= 8; \
+        bits |= 0xFF; \
+        llvk -= 8; \
+        *cdata = bits >> bit_count; \
+        cdata++; \
+    } \
+    bits <<= llvk; \
+    bits |= (1 << llvk) - 1; \
+    bit_count += llvk; \
+    /*OutputBit(1, 0);*/ \
+    bits <<= 1; \
+    bit_count++; \
+    CheckWrite; \
+    /* remainder part of GR code, if needed */ \
+    if (lkr) \
+    { \
+        /*OutputBits(lkr, _lmag & ((1 << lkr) - 1));*/ \
+        bits <<= lkr; \
+        bits |= _lmag & ((1 << lkr) - 1); \
+        bit_count += lkr; \
+    } \
+    /* update _krp, only if it is not equal to 1 */ \
+    if (lvk == 0) \
+    { \
+        /*UpdateParam(_krp, -2, lkr);*/ \
+        _krp -= 2; \
+        if (_krp < 0) \
+        { \
+            _krp = 0; \
+        } \
+    } \
+    else if (lvk > 1) \
+    { \
+        /*UpdateParam(_krp, lvk, lkr);*/ \
+        _krp += lvk; \
+        if (_krp > 80) \
+        { \
+            _krp = 80; \
+        } \
+    } \
+} while (0)
+
+int
+rfx_encode_diff_rlgr1(sint16 *coef, uint8 *cdata, int cdata_size)
+{
+    int k;
+    int kp;
+    int krp;
+
+    int input;
+    int numZeros;
+    int runmax;
+    int mag;
+    int sign;
+    int processed_size;
+    int lmag;
+    int coef_size;
+    int y;
+
+    int bit_count;
+    unsigned int bits;
+    uint8 *cdata_org;
+
+    uint32 twoMs;
+
+    for (k = 4095; k > 4032; k--)
+    {
+        coef[k] -= coef[k - 1];
+    }
+
+    //rfx_bitstream_attach(bs, cdata, cdata_size);
+
+    /* initialize the parameters */
+    k = 1;
+    kp = 1 << 3;
+    krp = 1 << 3;
+
+    bit_count = 0;
+    bits = 0;
+    cdata_org = cdata;
+
+    /* process all the input coefficients */
+    coef_size = 4096;
+    while (coef_size > 0)
+    {
+        if (k)
+        {
+
+            /* RUN-LENGTH MODE */
+
+            /* collect the run of zeros in the input stream */
+            numZeros = 0;
+
+            GetNextInput;
+            while (input == 0 && coef_size > 0)
+            {
+                numZeros++;
+                GetNextInput;
+            }
+
+            /* emit output zeros */
+            runmax = 1 << k;
+            while (numZeros >= runmax)
+            {
+
+                //OutputBit(1, 0); /* output a zero bit */
+                bits <<= 1;
+                bit_count++;
+
+                CheckWrite;
+
+                numZeros -= runmax;
+
+                //UpdateParam(kp, UP_GR, k); /* update kp, k */
+                kp += 4;
+                if (kp > 80)
+                {
+                    kp = 80;
+                }
```

Comment:
What is 80 here? Would it be possible to write `kp = max(kp, 80)`? It would be more readable.